### PR TITLE
Kernel/Graphics: Introduce the DisplayConnector design

### PR DESCRIPTION
**Pull request by supercomputer7**

### CODE REVIEW

**Comment by supercomputer7** [comment]
I'm not marking this as a draft, but I do work on "porting" the Intel driver to use this neat design too now. If this is a issue, please let me know :)

**Comment by supercomputer7** [comment]
The goal of this PR now is to establish organized API to modeset a display resolution like it's on a real hardware. Until this PR, we just tell the kernel to put an arbitrary width and height but on real hardware it's a lot more complicated (well, the EDID blob helps here), so this PR will change this behavior to help writing more native drivers hopefully :)

**Comment by supercomputer7** [comment]
I decided I should try to merge this as soon as possible, so it's not a draft anymore but another step before implementing bare metal modesetting (and I mean - allowing userspace to change resolutions on bare metal). This mainly applies to the Intel driver but other bare metal drivers in the future can use that pattern.
The VirtIO code was not touched on purpose - changing it requires lots of work both on implementing modesetting of the virtual hardware and also to put the code to be in the programming style of the rest of the repository.

I think in the overall result, this PR made very nice changes to the graphics subsystem but we definitely need more work and effort to put it on a certain level so we can use it on bare metal hardware too.

**Comment by awesomekling** [comment]
What's a DisplayConnector?

**Comment by supercomputer7** [comment]
> What's a DisplayConnector?

It's somewhat like a `drm_connector` struct on the linux kernel:
https://github.com/torvalds/linux/blob/master/include/drm/drm_connector.h

The idea was to build a class that handles all graphical operations on a single display connection as real hardware tends to have multiple sets of display connections (you can look on your x86 machine, you probably have a VGA port, a HDMI port and maybe DisplayPort port). I could call it `DisplayPort` but that will make people confused with the real standard and "connector" is also a good option.
I build this because I want to make the kernel to be able to modeset a resolution on bare metal from userspace - currently the Intel driver chooses the best resolution for you but you can't really change it from `DisplaySettings` which is a very nice feature. So, to do that, I want that `DisplaySettings` to hold a related data from the EDID blob so if the user chooses a certain resolution, then he sends the timings of that resolution instead of width and height as we do now. This is because a display monitor can have for example a resolution of 1024x768 for 60Hz and 1024x768 for 75Hz and the only sane way to deal with that is to use the LibEDID much more broadly in userspace and the kernel.

Do you want me to clarify this in the commit message when I introduce this class?

**Comment by supercomputer7** [comment]
I'd like to wait with merging this PR until #12209 is merged. That is because @tomuta put very useful changes in regard to changing the framebuffer console I'd want to rely on.

**Comment by tomuta** [comment]
> build a class that handles all graphical operations on a single display connection as real hardware tends to have multiple sets of display connections (you can look on your x86 machine, you probably have a VGA port, a HDMI port and maybe DisplayPort port). I could call it `DisplayPort` but that will make people confused with the real standard and "connector" is also a good option.
> I build this because I want to make the kernel to be able to modeset a resolution on bare metal from userspace - currently the Intel driver chooses the best resolution for you but

So how does this differ from the `FramebufferDevice` we have in concept? For example in VirtIO you have a `FramebufferDevice` for each physical output, and you can change resolutions on each independently. I imagined we'd do the same with any real hardware driver?

**Comment by tomuta** [comment]
> I'd like to wait with merging this PR until #12209 is merged. That is because @tomuta put very useful changes in regard to changing the framebuffer console I'd want to rely on.

This was merged yesterday.

**Comment by supercomputer7** [comment]
> > build a class that handles all graphical operations on a single display connection as real hardware tends to have multiple sets of display connections (you can look on your x86 machine, you probably have a VGA port, a HDMI port and maybe DisplayPort port). I could call it `DisplayPort` but that will make people confused with the real standard and "connector" is also a good option.
> > I build this because I want to make the kernel to be able to modeset a resolution on bare metal from userspace - currently the Intel driver chooses the best resolution for you but
> 
> So how does this differ from the `FramebufferDevice` we have in concept? For example in VirtIO you have a `FramebufferDevice` for each physical output, and you can change resolutions on each independently. I imagined we'd do the same with any real hardware driver?

If hardware was a straight forward subsystem then technically there's no different for this. However, for each GraphicsAdapter there's a framebuffer device (at least this is the direction I want it to be into) - we definitely don't want to spawn multiple framebuffer devices for multiple screens - think about what happens when we don't have nice hardware to give us a framebuffer but instead all you have is a MMIO region and slim memory region to put pixels and then you send commands to do DMA to the internal storage of the pixels on the card. As you might see right now, each `FramebufferDevice` can support multiple heads.

The whole idea of `DisplayConnector` is to allow one Graphics adapter to have multiple sets of this type of object but abstracted from its knowledge on how precisely each one works so if you have a modern Intel machine with multiple outputs then you might have a DisplayConnector for controlling VGA connector, another connector for HDMI and maybe even DisplayPort one which should also include functionality of controlling the AUX channel (which is an i2c-like bus?) and be able to train it...
I just really don't want bulky graphics adapter classes with lots of code nobody can read, and this class not only gives the opportunity to put pieces of code into logical objects & classes, I also want this to take inspiration from how the DRM subsystem works in Linux - in the best case scenario, this could lead to a strong abstraction layer of handling complicated graphics stuff and still to be written in a sane manner we can all read and hack on :)

**Comment by tomuta** [comment]
Ok so how does e.g. WindowServer know when a display was disconnected or added? A device for each physical output can be pulled by the kernel, no problem. Finding new ones by watching /dev is also easy. But if we're merely dealing with a head index like in master right now that seems not very flexible or easy to deal with. Looking at what you came up with I guess it's effectively just a rename from FramebufferDevice to DisplayConnector, right? WindowServer simply just openes a DisplayConnector and maps the framebuffer, right? Although your example with the mmio region and slim memory I think that still can be implemented just fine with our current approach, in fact VirtIO already is kind of like this. At any rate, was just curious about the thinking behind this. Tbh I don't know anything about how Linux implements this ;)

**Comment by supercomputer7** [comment]
> Ok so how does e.g. WindowServer know when a display was disconnected or added? A device for each physical output can be pulled by the kernel, no problem. Finding new ones by watching /dev is also easy. But if we're merely dealing with a head index like in master right now that seems not very flexible or easy to deal with. Looking at what you came up with I guess it's effectively just a rename from FramebufferDevice to DisplayConnector, right? WindowServer simply just openes a DisplayConnector and maps the framebuffer, right? Although your example with the mmio region and slim memory I think that still can be implemented just fine with our current approach, in fact VirtIO already is kind of like this. At any rate, was just curious about the thinking behind this. Tbh I don't know anything about how Linux implements this ;)

> Ok so how does e.g. WindowServer know when a display was disconnected or added?

Currently we don't know about that really, but in theory WindowServer should either poll the framebuffer device or let the `udev-like` daemon to remove the device from /dev entirely. Both solutions are incomplete and have their own issues.

> Looking at what you came up with I guess it's effectively just a rename from FramebufferDevice to DisplayConnector, right?

It's somewhat correct, but I rather define it as more like a shift in theory of operation, just like how we started with no devfs at all, then I introduced the devfs and then I removed it to create devtmpfs :)
We use framebuffer devices now, but real hardware is much more complicated today and framebuffer devices limit our capabilities to 2D rendering at best and create several issues we need to deal with today - for example, lack of sync with VT console switch in the kernel and letting WindowServer printing pixels on the same framebuffer from userspace (there are solutions to this, like deferred IO which Linux implemented).
So in some sense, we might expose our own "`/dev/dri`" directory in the future and the char devices there will refer to a `DisplayConnector` on the kernel :)

> Although your example with the mmio region and slim memory I think that still can be implemented just fine with our current approach, in fact VirtIO already is kind of like this. 

Currently the `DisplayConnector` is nothing more than a way to split modesetting code from the bulky graphics adapters code so it is not exposed to the userspace in any way. Yes, the way we do things now is fine (but see other notes I put here to describe why it's really not near an ideal solution) but we will need to consider this soon, especially since I know I'm not the only one eager to see us supporting modern hardware and 3D functionality in the OS, backed by hardware :)

Reference to what I talk about in general, maybe it will help resolve other questions:
https://en.wikipedia.org/wiki/Direct_Rendering_Manager#Kernel_Mode_Setting

**Comment by supercomputer7** [comment]
This is postponed until other PRs I want to merge are merged first.

**Comment by supercomputer7** [comment]
I'd like to get this merged so I can continue to work on this. For now, I upstream'd a commit to a different branch to let other developers see the direction of this:
https://github.com/supercomputer7/serenity/commit/8eeed48dfdd8bbe09694cfc76267da8b71bd9da5

File: Kernel/Graphics/VGA/PCIGenericAdapter.cpp
Comment (ccapitalK) [inline Kernel/Graphics/VGA/PCIGenericAdapter.cpp:38]: Are these meant to be different fields being checked?

File: Kernel/Graphics/VGA/GenericAdapter.h
Comment (ccapitalK) [inline Kernel/Graphics/VGA/GenericAdapter.h:2]: Why are you changing in the copyright years?

File: Kernel/Graphics/Intel/NativeDisplayConnector.h
Comment (ccapitalK) [inline Kernel/Graphics/Intel/NativeDisplayConnector.h:52]: Nit: Should be a blank line between these two structs.

File: Kernel/Graphics/Intel/NativeDisplayConnector.cpp
Comment (ccapitalK) [inline Kernel/Graphics/Intel/NativeDisplayConnector.cpp:24]: The brackets here are a bit strange. I believe this expression is parsed as `((refclock * ((5 * m1) + m2)) / n) / (p1 * p2)`, which is equivalent to `(refclock * (5 * m1 + m2) / n) / (p1 * p2)`. I'm assuming this calculation is what you intend, the brackets here around `m1`, `m2` and `n` are unnecessary.

File: Kernel/Graphics/Intel/NativeDisplayConnector.cpp
Comment (ccapitalK) [inline Kernel/Graphics/Intel/NativeDisplayConnector.cpp:29]: Same as above.

File: Kernel/Graphics/Intel/NativeDisplayConnector.cpp
Comment (ccapitalK) [inline Kernel/Graphics/Intel/NativeDisplayConnector.cpp:34]: Same as above.

**Review by ccapitalK — COMMENTED** [review]
Couple of nitpicks, other than that looks good to me.

**Comment by ccapitalK** [comment]
@supercomputer7 Uh, this PR is starting to become very large. For the sake of not making a monster PR I think it would be wise to hold off on adding the new VMWare SVGA adaptor driver in this one, just do what is necessary to port the existing adaptors to the new design and then push the SVGA code in another PR.

**Comment by supercomputer7** [comment]
> @supercomputer7 Uh, this PR is starting to become very large. For the sake of not making a monster PR I think it would be wise to hold off on adding the new VMWare SVGA adaptor driver in this one, just do what is necessary to port the existing adaptors to the new design and then push the SVGA code in another PR.

I agree, I could put this off in a different PR or just wait with it :)

**Comment by supercomputer7** [comment]
Let's get this merged, as I want to send a patch to add VMWare SVGA support too. I have nice patches locally now (they were on this branch but I dropped them so it's easier to review this). As always, I'd be happy to hear suggestions and comments about this PR :)

File: Kernel/TTY/TTY.cpp
Comment (supercomputer7) [inline Kernel/TTY/TTY.cpp:316]: Probably should not include this diff in this PR.

File: Userland/Services/SystemServer/main.cpp
Comment (supercomputer7) [inline Userland/Services/SystemServer/main.cpp:105]: Should delete this newline.

**Comment by ccapitalK** [comment]
> Let's get this merged, as I want to send a patch to add VMWare SVGA support too. I have nice patches locally now (they were on this branch but I dropped them so it's easier to review this). As always, I'd be happy to hear suggestions and comments about this PR :)

Alright, I'll take a look at this later today. I also want this merged, since I want to do some improvements to the VirtIOGPU adapter as well (make it async + improve the ioctl layout).

**Comment by supercomputer7** [comment]
> > Let's get this merged, as I want to send a patch to add VMWare SVGA support too. I have nice patches locally now (they were on this branch but I dropped them so it's easier to review this). As always, I'd be happy to hear suggestions and comments about this PR :)
> 
> Alright, I'll take a look at this later today. I also want this merged, since I want to do some improvements to the VirtIOGPU adapter as well (make it async + improve the ioctl layout).

I see. In general it feels like this design is more robust but still there's a lot of expansion to make this really "perfect". I started to work on an ATOMBIOS parser driver locally for the sake of supporting my Radeon HD5450 video card in SerenityOS. It really seems like going forward with this design, we can eventually have also 3D acceleration being fit really nicely into the OS.

**Comment by supercomputer7** [comment]
Rebased after moving to clang-format-14.

File: Kernel/Graphics/VirtIOGPU/Console.h
Comment (ccapitalK) [inline Kernel/Graphics/VirtIOGPU/Console.h:13]: Since you've put both VirtIODisplayConnector and VirtIOGraphicsAdapter directly in the Kernel namespace, it might be more consistent to name this console VirtIOConsole and put it in the Kernel namespace as well, instead of keeping it in `Kernel::Graphics::VirtIOGPU::Console`.

File: Kernel/Graphics/VirtIOGPU/DisplayConnector.cpp
Comment (ccapitalK) [inline Kernel/Graphics/VirtIOGPU/DisplayConnector.cpp:85]: Nit: This probably shouldn't return VirGL if VirGL support wasn't negotiated between the Graphics Adapter and the hardware.

File: Kernel/Graphics/VirtIOGPU/DisplayConnector.cpp
Comment (ccapitalK) [inline Kernel/Graphics/VirtIOGPU/DisplayConnector.cpp:309]: Nit: We probably add some way of preventing arbitrary userspace programs from writing to the framebuffer, if we haven't already.

**Comment by ccapitalK** [comment]
Once the label and namespace issues have been resolved, I'd say this is good to merge.

File: Kernel/Graphics/VirtIOGPU/DisplayConnector.cpp
Comment (supercomputer7) [reply Kernel/Graphics/VirtIOGPU/DisplayConnector.cpp:309]: We don't have a way to prevent it currently (in master too), so let's worry about this later on :)

File: Kernel/Graphics/DisplayConnector.h
Comment (supercomputer7) [reply Kernel/Graphics/DisplayConnector.h:79]: When we need to worry about non-contiguous apertures, we can re-write this interface. For now I think this fine :)

**Review by ccapitalK — APPROVED** [review]
LGTM

**Comment by supercomputer7** [comment]
Down the road, I'd like to get this merged because I have more nice patches to send after this.
After this being merged, there's the VMWare patch to add hardware support for VMWare SVGA II:
https://github.com/supercomputer7/serenity/tree/graphics-display-connectors-vmware
And then I polished the Intel driver here:
https://github.com/supercomputer7/serenity/tree/graphics-display-connectors-intel-focused
so it looks very clean and ready for future generations of the Intel graphics chipset.

Many nice features and fixes are already planned in a mainline subtree without any additional hardware support here:
https://github.com/supercomputer7/serenity/tree/graphics-display-connectors-mainline

**Comment by ccapitalK** [comment]
I think realistically the only way we are going to get this merged is to close it and split it into smaller PRs as a series. I would recommend linking this PR in the new ones to show there was some degree of review.

**Comment by supercomputer7** [comment]
> I think realistically the only way we are going to get this merged is to close it and split it into smaller PRs as a series. I would recommend linking this PR in the new ones to show there was some degree of review.

I'm not against this, but the maintainers are completely silent about this so I can't really tell what to do.

**Comment by supercomputer7** [comment]
@ccapitalK  Alright, let's try this your way then. I will send small PRs to allow merging of this instead of doing everything in one piece.

**Pull request closed without merging**
